crates/sui/src/upgrade_compatibility.rs

```diff
@@ -895,6 +896,36 @@ fn function_lost_public(
     Ok(diags)
 }
 
+fn format_param(
```

Comment:
FWIW, this function will behave more nicely when it comes to allocations if you structure it to accept a buffer to write into, rather than return a string:

```rust
fn format_params(
    param: &Type,
    type_params: Vec<SourceName>,
    output: &mut String,
    secondary: &mut Vec<(Loc, String)>,
) -> Result<(), Error>;
```

Instead of `format!(...)` you can use `write!(output, ...).unwrap()` -- it is safe to unwrap because writing to a string doesn't panic. With this pattern, you don't end up allocating small temporary strings and then adding them onto the larger main output string.